### DISCUSSION TITLE
JS SDK: Auto-cancel newer requests

### DISCUSSION THREAD

**Discussion by j0sh** [pocketbase/pocketbase#4454] [Q&A]
It would be nice to have an option to auto-cancel newer requests (eg, preventing them from being transmitted entirely) rather than cancelling older ones. The reason is that this only a client-side cancel; if there are side effects on the server from the first request, then the second call may fail or return unexpected results.

Concretely, this is happening with the `authWithOAuth2Code` call and the Github provider with manual code exchange flow. The OAuth2 code cannot be used twice, which often leads to failures in React strict mode when the first request is cancelled by the client (but succeeds on the server), while the second request fails on the server. (The all-in-one flow is not used here because SSE connections are unreliable; see https://github.com/pocketbase/pocketbase/discussions/3766)

If this is an option that would be welcome, I can try adding it to the JS SDK. If not, is there another workaround for this issue?

**Comment by ganigeorgiev** [comment]
I don't understand what do you mean by _"Auto-cancel newer requests"_.

> The OAuth2 code cannot be used twice, which often leads to failures in React strict mode when the first request is cancelled by the client (but succeeds on the server), while the second request fails on the server. 

We are not storing the OAuth2 code anywhere so the above doesn't make sense to me (what first and second requests?).

> If this is an option that would be welcome, I can try adding it to the JS SDK. If not, is there another workaround for this issue?

It is not clear to me what actually is the issue you are describing or trying to fix.

**Reply by j0sh** [reply]
In React's [strict mode](https://react.dev/reference/react/StrictMode) (the default during development), the page is rendered twice, including any JS calls that are done automatically when the page loads. So when we do [manual code exchange](https://pocketbase.io/docs/authentication/), the `authWithOAuth2Code` API gets called twice in a React development environment because it is meant to be called automatically when the redirect page loads. The first call gets cancelled on the client, while the second fails on the server.

Hope that helps.

**Reply by j0sh** [reply]
The proposal is to add an option to the JS SDK to cancel newer requests, rather than older ones:

```
pb.collection('example').authWithOAuth2Code(...) // executed
pb.collection('example').authWithOAuth2Code(...) // ignored
```

This should probably be the default behavior for non-imdepotent requests (eg, POSTs) but a manually specified option in the request would be okay too.

For now I am working around the issue by disabling auto-cancel for the `authWithOAuth2Code` call but that still leads to a server-side failure for the 2nd call (and in my case, it is a race condition; the second request usually fails first, leading to a error flash in the UI until the first request completes and janks things to normal)

**Reply by ganigeorgiev** [reply]
I'm sorry but I don't see how this is PocketBase related.

Keep in mind that not all API responses are idempotent and there is no guarantee that re-rendering everything twice will always return the same result (for example fetching a collection sorted with `@random`).

Additionally, I don't understand how do you expect the second call to be "ignored" - what would be the return result in the second call? It cannot be just no-op because if you have an assignment or some other check it will be required to handle the return result conditionally in your code (once with the `Promise` and another time with a `undefined`) which to me doesn't sound right. And if you want to reject following requests wouldn't that also lead to different results during the re-rendering?

**Reply by ganigeorgiev** [reply]
I don't have the full picture of your use case and haven't used React in ages, but a workaround for the re-rendering if you don't want to disable strict mode could be to use a flag (ex. with [`useRef`](https://react.dev/reference/react/useRef) or similar) and manually check whether you want to run the request the second time.

It is possible to add option to cancel/reject the `Promise`s of any following requests matching a `requestKey` but I think it will be confusing to the users and will have a very limited use case.

**Reply by j0sh** [reply]
> I'm sorry but I don't see how this is PocketBase related.

Pocketbase already does request auto-cancellation (presumably for similar reasons?). This would be a knob to make things behave a little better under certain circumstances.

> Additionally, I don't understand how do you expect the second call to be "ignored" - what would be the return result in the second call?

Some options:
* Request coalescing - clone the response from the first request
* Immediately throw a cancellation exception. This is probably preferable for consistency with the current behavior.
 
> And if you want to reject following requests wouldn't that also lead to different results during the pre-rendering?

Yes, I actually have wrapper logic before pretty much every Pocketbase call to ignore cancellations due to re-rendering, instead of flashing an error in the UI. Eg:
```js
     try {
        const authData = await pb.collection(PB_Collection).authWithOAuth2Code(...)
        ...
      } catch(e:unknown){
        if (isCancelledRequest(e)) {
        } else if (e instanceof Error) {
          setResponseError(e.message);
        } else {
          setResponseError("Unknown error:" +e);
        }
        return
 ```

> but a workaround for the re-rendering if you don't want to disable strict mode could be to use a flag (ex. with [useRef](https://react.dev/reference/react/useRef) or similar) and manually check whether you want to run the request the second time.

Yeah - I was hoping to address the root of the issue, but can probably work around it on a higher level. Thanks.
